Fix accidental building of child records when `inverse_of` is specified

clouvet opened the pull request:
### Summary

If an existing record with inverse_of children was duplicated, additional child records would be built. These new records could be persisted upon save or update inadvertently. I added a method to address that case and prevent the accidental creation of additional associations. 

I traced the problem back to [this line in the `replace_on_target` method](https://github.com/clouvet/rails/blob/ecd878bba181e9e89d523e20802f3f46b2d91431/activerecord/lib/active_record/associations/collection_association.rb#L469) in `CollectionAssociation`. Using the steps to reproduce the problem in the original issue as an example, I wrote a test to cover the case of new child records being added to existing owner records when inversing is in play. 

With this change in place, the original steps to reproduce the problem deliver the expected result of no additional records. 

Fixes #44819.

clouvet committed:
Fix accidental building of child records when `inverse_of` is specified

If an existing record with inverse_of children was duplicated,
additional child records would be built. These new records could
be persisted upon save or update inadvertently.

This adds a method to check that new records with inversing have an
existing owner:

```ruby
new_record_inverse_of_existing_owner?(record, inversing: false)
```

Fixes #44819.

clouvet commented:
I force-pushed an update here to resolve an `activerecord/CHANGELOG.md` conflict.

clouvet commented:
@ghiculescu, since it's been a while I'm giving this a little nudge. I'm happy to revise or close this pull request if this approach isn't preferred.

ghiculescu commented:
Thanks @clouvet, but this would need review from a member of the Rails core / committers teams. I think the approach is fine. They are pretty busy, but they will eventually get to it :)

It looks like you have a merge conflict btw.

clouvet commented:
Thanks, @ghiculescu, I appreciate the response. I resolved the conflict with `activerecord/CHANGELOG.md` again.

ghiculescu commented on activerecord/CHANGELOG.md:
````diff
@@ -1,3 +1,20 @@
+    additional child records would be built. These new records could
+    be persisted upon save or update inadvertently.
+
+    This adds a method to check that new records with inversing have an
+    existing owner:
+
+    ```ruby
+    new_record_inverse_of_existing_owner?(record, inversing: false)
+    ```
````
I don't think we need this in the changelog, as it's an implementation detail, and not something users will interact with.

ghiculescu commented on activerecord/lib/active_record/associations/collection_association.rb:
```diff
@@ -465,7 +465,7 @@ def replace_on_target(record, skip_callbacks, replace:, inversing: false)
           elsif @_was_loaded || !loaded?
             @association_ids = nil
-            target << record
+            target << record if !new_record_inverse_of_existing_owner?(record, inversing: inversing)
```
```suggestion
            target << record unless new_record_inverse_of_existing_owner?(record, inversing: inversing)
```

clouvet commented:
@ghiculescu I made changes here in response to your review: I dropped the CHANGELOG addition, accepted your suggestion to use `unless` instead of `if !`, and revised the comment on `new_record_inverse_of_existing_owner?` to explain why that new method exists. 

I'm happy to make additional changes if necessary, and of course I'll squash these multiple commits down to one when the time comes. 

Thanks for the review!

clouvet commented:
Sorry for the noise here. I squashed the cleanup commits to make this pull request suitable for merging.

eileencodes commented on activerecord/lib/active_record/associations/collection_association.rb:
```diff
@@ -475,6 +475,12 @@ def replace_on_target(record, skip_callbacks, replace:, inversing: false)
 
+        # This check prevents the accidental creation of duplicate child records
+        # when inverse_of is in use for an already existing owner.
+        def new_record_inverse_of_existing_owner?(record, inversing: false)
```
The name is a bit of a mouthful, how about `inverse_of_existing_owner`

```suggestion
        def inverse_of_existing_owner?(record, inversing: false)
```

Also, can you move this under the private namespace, we don't want applications calling this directly.

clouvet replied:
I renamed the method and moved it to the private namespace.

eileencodes commented on activerecord/test/cases/associations/inverse_associations_test.rb:
```diff
@@ -872,6 +872,16 @@ def test_building_has_many_parent_association_inverses_one_record
     end
   end
+
+  def test_dup_with_inversing_does_not_create_additional_records_when_owner_already_exists
```
Do you know if there are tests for cases where one of the three `record.new_record? && inversing && !owner.new_record?` isn't true? If not we should add those as well.

clouvet replied:
I added more tests that have hopefully more descriptive names to cover cases for when one of `record.new_record? && inversing && !owner.new_record?` is not true. I didn't see existing tests that covered those situations.

The new tests point to a problem though. My proposed solution of checking `inverse_of_existing_owner?` breaks inversing for unsaved objects. This test demonstrates that:
```ruby
  def test_existing_parent_and_new_child_with_inversing_does_not_create_additional_associations
    with_has_many_inversing(Interest) do
      human = Human.create!
      interest = Interest.new(human: human)
      human.update!(name: "name")
      assert_equal human, interest.human
      assert_equal 0, human.interests.size
    end
  end
```

On the current `main` branch, `human.interests.size` should be `1` because of the inversing between the models. So my proposed solution is no good. 

I went around in circles on this point and along the way discovered that `inverse_of` results in autosaving of associated objects. It's unclear to me if it's supposed to, though I looked through the documentation for clarity on that point. 

I pushed up the new commits to show the process I went through in case it's beneficial to solving the original issue, but I imagine this pull request can be closed.

clouvet committed:
Change method name to inverse_of_existing_owner

And move to private namespace.

clouvet committed:
Add additional tests to cover the conditions of inverse_of_existing_owner?

These tests cover the different possible conditions of
`inverse_of_existing_owner?`:

`record.new_record? && inversing && !owner.new_record?`

Unfortunately, they highlight a problem with this solution. An
existing parent record with a new child association and inversing
breaks the inversing feature for unsaved records. For example:

```
  def test_existing_parent_and_new_child_with_inversing_does_not_create_additional_associations
    with_has_many_inversing(Interest) do
      human = Human.create!
      interest = Interest.new(human: human)
      human.update!(name: "name")
      assert_equal 0, human.interests.size
    end
  end
```

In `main`, `human.interests.size` equals `1`.
`inverse_of_exising_owner?` breaks that.

clouvet committed:
Remove unnecessary tests

Also adjust tests for the linter.

clouvet committed:
Remove extra lines at end of class
